### PR TITLE
Create local account

### PR DESCRIPTION
**Pull request by opixelum**
- Local account is needed for using the app
- App manages only 1 local account
- Local account is just a password (no username needed)
- Local account password is stored is `local_account` file at the root of the app directory
- This file contains the salt used for local account creation & the sha512 hash of the password
- Manage local account added in main menu
- Manage local account menu is for changing local account password and for deleting local account

### CODE REVIEW

File: lib/local_account.c [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    // Open local account file
+    FILE *local_account_file = fopen("local_account", "w");
+    if (!local_account_file)
+    {
+        fprintf(stderr, "Failed to open local account file.\n");
+        return -2;
+    }
+
+    // Write salt & password hash hex to local account file on different lines
+    fprintf(local_account_file, "%s\n", salt);
```
Comment (Invorom) [inline lib/local_account.c:79]: Need to encrypt salt

File: lib/local_account.c [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    // Open local account file
+    FILE *local_account_file = fopen("local_account", "r");
+    if (!local_account_file)
+    {
+        fprintf(stderr, "Failed to open local account file.\n");
+        return -2;
+    }
+
+    // Read salt from local account file
+    char *salt = malloc(sizeof *salt * SALT_LENGTH);
```
Comment (Invorom) [inline lib/local_account.c:105]: Change this if salt is encrypted

File: lib/local_account.c [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+    {
+        fprintf(stderr, "Memory allocation for password hash hex failed.\n");
+        return -2;
+    }
+    for (int i = 0; i < SHA512_DIGEST_LENGTH; i++)
+        sprintf(password_hash_from_user_hex + (i * 2), "%02x", password_hash_from_user[i]);
+    free(password_hash_from_user);
+    password_hash_from_user_hex[SHA512_DIGEST_LENGTH * 2] = '\0'; // Add null terminator
+
+    // Compare password hashes
```
Comment (Invorom) [inline lib/local_account.c:167]: Need to precise hexa

File: lib/ui.c [hunk 4]
```diff
@@ -1,17 +1,156 @@
 #include "ui.h"
 
-void mainMenu(unsigned char *isRunning)
+unsigned char getCharHide()
```
Comment (Invorom) [inline lib/ui.c:3]: U need to comment

File: lib/ui.c [hunk 5]
```diff
@@ -1,17 +1,156 @@
...
+    return (unsigned char)ch;
+}
+
+char *getStringHide()
+{
+    char *string = calloc(1024, sizeof *string);
+    if (!string)
+    {
+        fprintf(stderr, "Couldn't allocate memory for buffer\n");
+        exit(EXIT_FAILURE);
```
Comment (Invorom) [inline lib/ui.c:49]: Need to review errors management

File: lib/ui.c [hunk 6]
```diff
@@ -1,17 +1,156 @@
...
+    }
+    string[0] = '\0';
+
+    unsigned char c;
+    unsigned char string_length = 0;
+
+    do
+    {
+        c = getCharHide();
```
Comment (Invorom) [inline lib/ui.c]: Change variable name

File: lib/ui.c [hunk 7]
```diff
@@ -1,17 +1,156 @@
...
+
+        if (c == '\n') break;
+        else if (c == 127 || c == 8) // Handle del & backspace
```
Comment (Invorom) [inline lib/ui.c]: Del: right --> left
Suppr: left --> right

File: lib/ui.c [hunk 8]
```diff
@@ -1,17 +1,156 @@
...
+            }
+        }
+        else
+        {
+            string[string_length] = c;
+            string_length++;
+            printf("*");
+        }
+    }
+    while (string_length < 1024);
```
Comment (Invorom) [inline lib/ui.c:80]: BONUS: Block user from send more than 1024
Example: Red, or vibrations when the length is reached

File: lib/ui.c [hunk 9]
```diff
@@ -1,17 +1,156 @@
...
+    // Reallocate memory to fit the string
+    string = realloc(string, sizeof string * string_length);
+    string[string_length] = '\0';
+
+    return string;
+}
+
+void clearStdin()
+{
+    int c;
```
Comment (Invorom) [inline lib/ui.c]: Variable name

File: lib/ui.c [hunk 10]
```diff
@@ -1,17 +1,156 @@
...
+    set_password:
+    printf("Enter a password: ");
+    do
+    {
+        password = getStringHide();
+
+        // Check if password respects the minimum requirements
+        if (!minimumPasswordRequirementsCheck(password))
+        {
+            printf("\nPassword doesn't respect the minimum requirements:\n- Length >= 12\n- At least one uppercase\n- At least one lowercase\n- At least one digit\n- At least one special character.\nPlease enter a valid password: ");
```
Comment (Invorom) [inline lib/ui.c:111]: BONUS: Need to precise error type
User-friendly :)

File: lib/ui.c [hunk 11]
```diff
@@ -1,17 +1,156 @@
...
+        printf("Enter your password: ");
+        password = getStringHide();
+
+        // Check if password is correct
+        local_account_login_status = connectLocalAccount(password);
+        if (local_account_login_status == -1) printf("\nPassword is incorrect.\n");
+        else if (local_account_login_status == -2) fprintf(stderr, "\nError while connecting to local account.\n");
+        free(password);
+    }
+    while (local_account_login_status != 0);
```
Comment (Invorom) [inline lib/ui.c:159]: BONUS: Implement blocked time if wrong

File: lib/ui.c [hunk 12]
```diff
@@ -1,17 +1,156 @@
...
+
+
```
Comment (Invorom) [inline lib/ui.c:160]: Delete this ugly useless line

File: lib/ui.c [hunk 13]
```diff
@@ -116,3 +260,131 @@ void createCredentialsDialogue()
...
+            printf("Invalid choice");
+            break;
+    }
+}
+
+void changeLocalAccountPasswordDialogue()
+{
+    printf("\nChange local account password\n\n");
+
+    char is_current_password_correct = 0;
```
Comment (Invorom) [inline lib/ui.c:322]: BONUS: Add "Forgot my password" --> Need to save user email

Comment (Invorom) [reply lib/ui.c:322]: And modify email in manage locale account

File: lib/ui.c [hunk 14]
```diff
@@ -116,3 +260,131 @@ void createCredentialsDialogue()
...
+
+        goto set_new_password;
+    }
+    free(new_password_confirmation);
+
+    // Create local account
+    if (createLocalAccount(new_password) == 0) printf("\nPassword changed successfully.\n");
+    else fprintf(stderr, "\nError while changing local account password\n");
+    free(new_password);
+}
```
Comment (Invorom) [inline lib/ui.c:380]: BONUS: Refacto code and use only one function for change and create password

File: lib/ui.c [hunk 15]
```diff
@@ -116,3 +260,131 @@ void createCredentialsDialogue()
...
+
+void deleteLocalAccountDialogue(unsigned char *isRunning, unsigned char *isConnected)
+{
+    printf("\nDelete local account\n\nPLEASE NOTE THAT THIS ACTION WILL DELETE ALL YOUR CREDENTIALS AND IS IRREVERSIBLE.\n\n");
```
Comment (Invorom) [inline lib/ui.c:384]: /!\

File: src/main.c [hunk 16]
```diff
@@ -1,15 +1,21 @@
-#include "password.h"
 #include "ui.h"
-#include "encdec.h"
 #include <unistd.h>
```
Comment (Invorom) [inline src/main.c]: Idk, need to test

File: src/main.c [hunk 17]
```diff
@@ -1,15 +1,21 @@
...
 
     while (isRunning)
     {
-        mainMenu(&isRunning);
+        if (!isConnected)
+        {
+            isLocalAccountExists()
+            ? localAccountLogInDialogue()
+            : localAccountCreationDialogue();
+            isConnected = 1;
```
Comment (Invorom) [inline src/main.c:15]: Error management

**Review by Invorom — CHANGES_REQUESTED** [review]

**Review by Invorom — CHANGES_REQUESTED** [review]
BUG: - Create local account
                - Press enter without text
                - Type something different than digits, letters; specials (del, arrows, f12, ...)

File: lib/local_account.c (hunk 1 above)
Comment (opixelum) [reply lib/local_account.c:79]: #37

File: lib/local_account.c (hunk 2 above)
Comment (opixelum) [reply lib/local_account.c:105]: Will be done in #37

File: lib/local_account.c (hunk 3 above)
Comment (opixelum) [reply lib/local_account.c:167]: Done in commit 4a0f70f

File: lib/ui.c (hunk 4 above)
Comment (opixelum) [reply lib/ui.c:3]: Done in commit a57eed1

File: lib/ui.c (hunk 5 above)
Comment (opixelum) [reply lib/ui.c:49]: Will be done in #33

**Review by Invorom — APPROVED** [review]

**Pull request merged**